src/lib/y2partitioner/dialogs/partition_size.rb

```diff
@@ -10,143 +13,335 @@ module Dialogs
+          @regions = regions
+          @largest_region = @regions.max_by(&:size)
+        end
+
+        # @macro seeAbstractWidget
+        def label
+          _("New Partition Size")
+        end
 
-          # FIXME: how much should we support multiple slots?
-          # Maximum size: largest slot
-          # Custom size: smallest slot possible
-          # Custom region: inside any slot
-          @slot = slots.first
-          r = @slot.region
-          # should Region have a #size?
+        def items
+          max_size_label = Yast::Builtins.sformat(_("Maximum Size (%1)"),
+            @largest_region.size.human_floor)
+          [
+            [:max_size, max_size_label],
+            [:custom_size, _("Custom Size")],
+            [:custom_region, _("Custom Region")]
+          ]
+        end
 
-          @max_size = r.block_size * r.length
+        def widgets
+          @widgets ||= [
+            MaxSizeDummy.new(@largest_region),
+            CustomSizeInput.new(@ptemplate, @regions),
+            CustomRegion.new(@ptemplate, @regions)
+          ]
         end
 
+        # @macro seeAbstractWidget
         def init
-          Yast::UI.ChangeWidget(Id(:max_size), :Enabled, false)
-          Yast::UI.ChangeWidget(Id(:manual_size), :Enabled, false)
+          self.value = (@ptemplate.size_choice ||= :max_size)
+          # trigger disabling the other subwidgets
+          handle("ID" => value)
         end
 
+        # @macro seeAbstractWidget
         def store
-          start_block = Yast::UI.QueryWidget(Id(:start_block), :Value)
-          end_block = Yast::UI.QueryWidget(Id(:end_block), :Value)
-          len = end_block - start_block + 1
-          bsize = @slot.region.block_size # where does this come from?
-          region = Y2Storage::Region.create(start_block, len, bsize)
-          @ptemplate.region = region
-        end
-
-        def max_size_term
-          term(
-            :LeftRadioButton,
-            Id(:max_size),
-            Opt(:notify),
-            # radio button text, %1 is replaced by size
-            Builtins.sformat(
-              _("Maximum Size (%1)"),
-              @max_size.to_human_string
-            )
-          )
+          w = current_widget
+          w.store
+          @ptemplate.region = w.region
+          @ptemplate.size_choice = value
         end
+      end
 
-        def manual_size_term
-          # radio button text
-          term(
-            :LeftRadioButtonWithAttachment,
-            Id(:manual_size),
-            Opt(:notify),
-            _("Custom Size"),
-            VBox(
-              Id(:manual_size_attachment),
-              MinWidth(
-                15,
-                InputField(Id(:size_input), Opt(:shrinkable), _("Size"))
-              )
-            )
-          )
+      # An invisible widget that knows a Region
+      class MaxSizeDummy < CWM::Empty
+        attr_reader :region
+
+        def initialize(region)
+          @region = region
```

Comment:
jsut note, it is not so visible, but empty have by default one param, which is widget_id. Here it is set to region, which is discutable as I am not sure how well it will work to have id that is not string.